File: code/python/mr-nukleo/mr_nukleo/hardware.py
from machine import Pin, ADC, PWM
import time
from config import Config


class NMRHardware:
    def __init__(self):
        # Initialize pins
        # TODO: Do we need to pull-up/down some of them?
        self.polarize_control = Pin(Config.POLARIZE_CONTROL_PIN, Pin.OUT, value=0)
        self.tx_positive = Pin(Config.TX_POSITIVE_PIN, Pin.OUT, value=0)
        self.tx_negative = Pin(Config.TX_NEGATIVE_PIN, Pin.OUT, value=0)
        self.tx_enable = Pin(Config.TX_ENABLE_PIN, Pin.OUT, value=0)
        self.rx_enable = Pin(Config.RX_ENABLE_PIN, Pin.OUT, value=0)
        self.adc = ADC(Config.ADC_PIN)

        # Ensure safe initial state
        self.set_polarize(0)
        self.switch_to_receive()

    def set_polarize(self, state):
        """Control the polarization coil."""
        self.polarize_control.value(state)

    def turn_off_polarize_adiabatic(self):
        """Turn off polarization coil adiabatically (fancy word for 'slowly')."""
        self.polarize_control.value(0)
        # Current will decay exponentially according to the circuit time constant
        # Allow time for adiabatic field rotation
        time.sleep(0.01)

    def switch_to_transmit(self):
        """Switch to transmit mode."""
        self.rx_enable.value(0)
        time.sleep(0.001)  # Small delay to ensure relays don't switch simultaneously
        self.tx_enable.value(1)

    def switch_to_receive(self):
        """Switch to receive mode."""
        self.tx_enable.value(0)
        self.tx_positive.value(0)
        self.tx_negative.value(0)
        time.sleep(0.001)  # Small delay to ensure relays don't switch simultaneously
        self.rx_enable.value(1)

    def read_adc(self):
        """Read ADC value."""
        return self.adc.read_u16()

File: code/cpp/mr_nucleo/src/hardware.cpp
#include "hardware.h"
#include "pico/time.h"
#include <iostream>

// Simplified Pin and ADC simulation for this example
// In a real implementation, these would interface with hardware
class Pin {
public:
    Pin(int pin, bool is_output, bool initial_state = false) 
        : pin_number(pin), output(is_output), state(initial_state) {
        std::cout << "Pin " << pin_number << " initialized as " 
                  << (output ? "output" : "input") << std::endl;
    }
    
    void value(bool new_state) {
        state = new_state;
        std::cout << "Pin " << pin_number << " set to " << state << std::endl;
    }
    
private:
    int pin_number;
    bool output;
    bool state;
};

class ADC {
public:
    ADC(int pin) : pin_number(pin) {
        std::cout << "ADC initialized on pin " << pin_number << std::endl;
    }
    
    uint16_t read_u16() {
        // Simulate reading an ADC value (random in this case)
        return rand() % 65536;
    }
    
private:
    int pin_number;
};

NMRHardware::NMRHardware() {
    // Initialize pins
    polarize_control = std::make_unique<Pin>(Config::POLARIZE_CONTROL_PIN, true, false);
    tx_positive = std::make_unique<Pin>(Config::TX_POSITIVE_PIN, true, false);
    tx_negative = std::make_unique<Pin>(Config::TX_NEGATIVE_PIN, true, false);
    tx_enable = std::make_unique<Pin>(Config::TX_ENABLE_PIN, true, false);
    rx_enable = std::make_unique<Pin>(Config::RX_ENABLE_PIN, true, false);
    adc = std::make_unique<ADC>(Config::ADC_PIN);
    
    // Ensure safe initial state
    set_polarize(false);
    switch_to_receive();
}

NMRHardware::~NMRHardware() = default;

void NMRHardware::set_polarize(bool state) {
    polarize_control->value(state);
}

void NMRHardware::turn_off_polarize_adiabatic() {
    polarize_control->value(false);
    // Current will decay exponentially according to the circuit time constant
    // Allow time for adiabatic field rotation
    sleep_ms(10);
}

void NMRHardware::switch_to_transmit() {
    rx_enable->value(false);
    sleep_ms(1);  // Small delay to ensure relays don't switch simultaneously
    tx_enable->value(true);
}

void NMRHardware::switch_to_receive() {
    tx_enable->value(false);
    tx_positive->value(false);
    tx_negative->value(false);
    sleep_ms(1);  // Small delay to ensure relays don't switch simultaneously
    rx_enable->value(true);
}

uint16_t NMRHardware::read_adc() {
    return adc->read_u16();
}

void NMRHardware::set_tx_positive(bool state) {
    tx_positive->value(state);
}

void NMRHardware::set_tx_negative(bool state) {
    tx_negative->value(state);
}
